cmd/agent/api/server.go

```diff
@@ -69,124 +68,64 @@ func StartServer(
+	if apiConfigEnabled {
+		extraHosts = append(extraHosts, apiConfigHostPort)
 	}
 
-	s := grpc.NewServer(opts...)
-	pb.RegisterAgentServer(s, &server{})
-	pb.RegisterAgentSecureServer(s, &serverSecure{
-		configService: configService,
-		taggerServer:  taggerserver.NewServer(tagger.GetDefaultTagger()),
-		// TODO(components): decide if workloadmetaServer should be componentized itself
-		workloadmetaServer: workloadmetaServer.NewServer(wmeta),
-		dogstatsdServer:    dogstatsdServer,
-		capture:            capture,
-	})
-
-	dcreds := credentials.NewTLS(&tls.Config{
-		ServerName: tlsAddr,
-		RootCAs:    tlsCertPool,
-	})
-	dopts := []grpc.DialOption{grpc.WithTransportCredentials(dcreds)}
-
-	// starting grpc gateway
-	ctx := context.Background()
-	gwmux := runtime.NewServeMux()
-	err = pb.RegisterAgentHandlerFromEndpoint(
-		ctx, gwmux, tlsAddr, dopts)
+	tlsKeyPair, tlsCertPool, err := initializeTLS(extraHosts...)
 	if err != nil {
-		return fmt.Errorf("error registering agent handler from endpoint %s: %v", tlsAddr, err)
+		return fmt.Errorf("unable to initialize TLS: %v", err)
 	}
 
-	err = pb.RegisterAgentSecureHandlerFromEndpoint(
-		ctx, gwmux, tlsAddr, dopts)
-	if err != nil {
-		return fmt.Errorf("error registering agent secure handler from endpoint %s: %v", tlsAddr, err)
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{*tlsKeyPair},
+		NextProtos:   []string{"h2"},
+		MinVersion:   tls.VersionTLS12,
 	}
 
-	// Setup multiplexer
-	// create the REST HTTP router
-	agentMux := gorilla.NewRouter()
-	checkMux := gorilla.NewRouter()
-	// Validate token for every request
-	agentMux.Use(validateToken)
-	checkMux.Use(validateToken)
-
-	mux := http.NewServeMux()
-	mux.Handle(
-		"/agent/",
-		http.StripPrefix("/agent",
-			agent.SetupHandlers(
-				agentMux,
-				flare,
-				dogstatsdServer,
-				serverDebug,
-				wmeta,
-				logsAgent,
-				senderManager,
-				hostMetadata,
-				invAgent,
-				demux,
-				invHost,
-				secretResolver,
-			)))
-	mux.Handle("/check/", http.StripPrefix("/check", check.SetupHandlers(checkMux)))
-	mux.Handle("/", gwmux)
-
-	// Use a stack depth of 4 on top of the default one to get a relevant filename in the stdlib
-	logWriter, _ := config.NewLogWriter(5, seelog.ErrorLvl)
-
-	srv := grpcutil.NewMuxedGRPCServer(
-		tlsAddr,
-		&tls.Config{
-			Certificates: []tls.Certificate{*tlsKeyPair},
-			NextProtos:   []string{"h2"},
-			MinVersion:   tls.VersionTLS12,
-		},
-		s,
-		grpcutil.TimeoutHandlerFunc(mux, time.Duration(config.Datadog.GetInt64("server_timeout"))*time.Second),
-	)
+	if err := util.CreateAndSetAuthToken(); err != nil {
+		return err
+	}
 
-	srv.ErrorLog = stdLog.New(logWriter, "Error from the agent http API server: ", 0) // log errors to seelog
+	if err := startCMDServer(
+		apiAddr, tlsConfig, tlsCertPool,
+		configService, flare, dogstatsdServer,
+		capture, serverDebug, wmeta, logsAgent,
+		senderManager, hostMetadata, invAgent,
+		demux, invHost, secretResolver,
```

Comment:
It may not be a convention in Go, but I see having too many parameters/arguments in a function as a code smell. 

Maybe a suggestion for the CMD server could be its struct that we initialize and later call `startServer` or something. I also understand it might not be the right suggestion for this particular use case. At least we can start a conversation 😄

Reply:
I definitely agree, IMO the API should at some point become a generic components others could register endpoints into, but this would be part of an eventual refactor of the API component

Reply:
FWIW, all these arguments will be removed in the near future as we'll embed them in the API Component directly (they will be injected to the API component through Fx)